File: mathlib/wasm-demo/camera/camera.js
/**
 * mathlib WASM demo — Camera matrices.
 */
import {
  initLib, byId, showError, needBuild,
  bindExampleSelector, renderMatrix4x4Float,
} from "../shared.js";

try {
  const lib = await initLib();
  const { WasmCg } = lib;

  const CAMERA_EXAMPLES = [
    { aspect: 16 / 9, fovY: Math.PI / 4, near: 0.1, far: 100 },
    { aspect: 1, fovY: Math.PI / 6, near: 0.5, far: 50 },
  ];
  const cameraResults = CAMERA_EXAMPLES.map((ex) => {
    const persp = WasmCg.newPerspective(ex.aspect, ex.fovY, ex.near, ex.far);
    const lookAt = WasmCg.lookAtRh(0, 0, 5, 0, 0, 0, 0, 1, 0);
    return { ...ex, persp, lookAt };
  });
  bindExampleSelector("camera-examples", ["Example 1", "Example 2"], (i) => {
    const r = cameraResults[i];
    byId("out-camera").innerHTML =
      "<strong>Perspective</strong> (aspect=" + r.aspect.toFixed(2) + ", fov=" + (r.fovY === Math.PI / 4 ? "π/4" : "π/6") + ", near=" + r.near + ", far=" + r.far + "):" +
      renderMatrix4x4Float(r.persp.toArray()) +
      "<strong>Look-at RH</strong> (eye 0,0,5 → target 0,0,0, up 0,1,0):" +
      renderMatrix4x4Float(r.lookAt.toArray());
  });
  byId("out-camera").innerHTML =
    "<strong>Perspective</strong> (aspect=1.78, fov=π/4, near=0.1, far=100):" +
    renderMatrix4x4Float(cameraResults[0].persp.toArray()) +
    "<strong>Look-at RH</strong> (eye 0,0,5 → target 0,0,0, up 0,1,0):" +
    renderMatrix4x4Float(cameraResults[0].lookAt.toArray());
} catch (e) {
  const out = byId("out-camera");
  if (out) { out.className = "error"; out.textContent = "Error: " + (e.message || String(e)); }
  showError((e.message || "").toLowerCase().includes("fetch") || (e.message || "").toLowerCase().includes("import") ? needBuild + "\n\n" : "" + (e.message || String(e)));
}

File: mathlib/tests/wasm/camera.rs
//! Integration tests for wasm camera bindings (demo: Camera matrices).

#![cfg(feature = "wasm")]

use mathlib::wasm::WasmCg;

#[test]
fn wasm_demo_camera() {
    // Demo: perspective (aspect=16/9, fov=π/4, near=0.1, far=100), look-at RH (eye 0,0,5 → target 0,0,0, up 0,1,0)
    let aspect = 16.0_f32 / 9.0;
    let fov_y = std::f32::consts::FRAC_PI_4;
    let persp = WasmCg::new_perspective(aspect, fov_y, 0.1, 100.0);
    assert_eq!(persp.rows(), 4);
    assert_eq!(persp.cols(), 4);
    let p = persp.to_array();
    assert!(p[0] > 1.0 && p[5] > 2.0);
    assert!(p[10] < 0.0);
    let look_at = WasmCg::look_at_rh(0.0, 0.0, 5.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0);
    assert_eq!(look_at.rows(), 4);
    assert_eq!(look_at.cols(), 4);
    let la = look_at.to_array();
    assert!((la[14] + 5.0).abs() < 0.01);
}
